pkg/apis/options/login_url_parameters.go

```diff
@@ -0,0 +1,54 @@
+package options
+
+import "regexp"
+
+// LoginURLParameter is the configuration for a single query parameter that
+// can be passed through from the `/oauth2/start` endpoint to the IdP login
+// URL.  The "default" option specifies the default value or values (if any)
+// that will be passed to the IdP for this parameter, and "allowed" is a
+// regular expression specifying which values (if any) that a caller passes
+// in the query string to `/oauth2/start` are permitted to override the defaults.
+// If _only_ a default is specified and no "allowed" then the parameter is
+// effectively fixed - the default value will always be used and anything
+// passed to the start URL will be ignored.  If _only_ "allowed" is specified
+// but no default then the parameter will only be passed on to the IdP if
+// the caller provides it, and no value will be sent otherwise.  Note that
+// the "allowed" regex is implicitly anchored with a `^` at the start and `$`
+// at the end, so it must match the _whole_ of the value to be considered
+// valid.
+type LoginURLParameter struct {
+	// Default specifies a default value or values that will be
+	// passed to the IdP if not overridden.
+	Default []string `json:"default,omitempty"`
+
+	// Allowed specifies a regular expression to validate any
+	// override values passed to the start endpoint.
+	Allowed *URLParameterRegexp `json:"allowed,omitempty"`
```

Comment:
Would it be easier if this was just a list of strings that are allowed?
Alternatively, is a list of regexes useful?

Reply:
I went through several rounds of changing my mind on this one.  The original version when I first created this PR was to have a list of allowed strings, but then I needed a separate "allow anything" flag to allow for use cases like #1369 that involve permitting any email address as a parameter value.

Given what you've said in other comments about keeping the `options` structs pure and mapping to the internal form in `ProviderData`, and also avoiding maps in the options yaml, maybe we should re-think the whole structure and go back to something more like my original design.  I'm thinking something like:

```yaml
loginURLParameters:
- name: prompt
  default: ["login"]
  allow:
  - value: consent
  - value: select_account
- name: login_hint
  allow:
  - pattern: .*@.*
- name: organization
  allow:
  - pattern: .*
```

i.e.

```go
type LoginUrlParameter struct {
        Name string, // required
        Default []string, // optional
        Allow []AllowedParameterValue, // optional
}

type AllowedParameterValue struct {
        Value string,
        Pattern string,
}
```

then in `ProviderData` the whole allow list (any mixture of fixed string values or patterns) could compile down to a single `*regex.Regex` - each `Value` becomes `"^(?:" + regexp.QuoteMeta(pv.Value) + ")$"`, each `Pattern` becomes `"(?:" + pv.Pattern + ")"` and then the whole lot can be joined with `|`.

I think I prefer this idea of multiple values/patterns and making literal value matches explicit, particularly if you want _not_ to anchor each pattern by default.  If you're not really familiar with regexes it's too easy to accidentally say `^foo|bar|baz$` (stats with foo, or contains bar, or ends with baz) when you meant `^(?:foo|bar|baz)$` (is exactly one of the three options).  Or am I over-thinking it?

Reply:
From an API design perspective, I like the sound of this latest suggestion. For the fact that it can all become one regex as well is quite convenient, let's go down that route for now then

Reply:
This also gets rid of the regex custom unmarshaller mess in options.